src/libraries/System.Linq/src/System/Linq/Cast.SpeedOpt.cs

```diff
@@ -3,11 +3,175 @@
 
 using System.Collections;
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
 
 namespace System.Linq
 {
     public static partial class Enumerable
     {
+        private sealed partial class OfTypeIterator<TResult>
```

Comment:
Nit: does the added logic justify extracting to separate `OfType*.cs` files?

Reply:
If you'd prefer that, I can do so, or we can do it subsequently.

Reply:
I think keeping it separate would be nice to have. The methods are similar but have distinct semantics and AFAICT they don't share logic.